### PR TITLE
Highlights share

### PR DESCRIPTION
**Pull request by uroybd**
This PR enables users to share exported highlights on Android.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9153)

### CODE REVIEW

**Comment by pazos** [comment]
@uroybd: made a few changes.

Mostly:
 
- don't mix logic. Use `canShareText` as a precondition for `doShareText`. Avoid specific android checks.
- disabled share submenu if there's no document opened (like the export one).

**Comment by uroybd** [comment]
> @uroybd: made a few changes.
> 
> Mostly:
>  
> - don't mix logic. Use `canShareText` as a precondition for `doShareText`. Avoid specific android checks.
> - disabled share submenu if there's no document opened (like the export one).
> 

It looks much cleaner now. I've also made the changes you've asked.

Should I add the same functionality for text, JSON, and html?

**Comment by uroybd** [comment]
@pazos 

Also, I need some help from luajit part.

The shareText function can do with two modifications:

1. Optional `EXTRA_TITLE` field. Some apps might use it to set the note/file name. We can send a generated title using this.

2. Option to set mimetype, defaults to `text/plain`.

**Comment by pazos** [comment]
> @pazos
> 
> Also, I need some help from luajit part.
> 
> The `doShareText` function can do with two modifications:
> 
> 1. Optional `EXTRA_TITLE` field. Some apps might use it to set the note/file name. We can send a generated title using this.
> 2. Option to set mimetype, defaults to `text/plain`.

Both are doable but not for this PR.  
They will require some familiary with https://github.com/koreader/android-luajit-launcher. 

Also the signature of share text is different https://github.com/koreader/android-luajit-launcher/blob/master/app/src/main/java/org/koreader/launcher/extensions/ActivityExtensions.kt#L258, because it was designed to deliver text to specific packages using specific actions, so something like that requires a reestructure I'm not willing to do in short run.

**Comment by pazos** [comment]
> Should I add the same functionality for text, JSON, and html?

Sure, go ahead.

Also you can test them on the emulator with

```diff
diff --git a/frontend/device/sdl/device.lua b/frontend/device/sdl/device.lua
index cef65e08..8b86498a 100644
--- a/frontend/device/sdl/device.lua
+++ b/frontend/device/sdl/device.lua
@@ -122,6 +122,10 @@ local Emulator = Device:new{
     canPowerOff = yes,
     canReboot = yes,
     canSuspend = yes,
+    canShareText = yes,
+    doShareText = function(self, text)
+        logger.info("sharing text", text)
+    end,
 }
```

**Comment by uroybd** [comment]
@pazos 

Instead of `isDocReady` I'm using `isUIReady` which doesn't check for enabled formats. Enabling formats are required for file generation. However, for on-off sharing directly selecting the format, it is not necessary.

**Comment by uroybd** [comment]
Removed WIP.

**Comment by pazos** [comment]
> @pazos
> 
> Instead of `isDocReady` I'm using `isUIReady` which doesn't check for enabled formats. Enabling formats are required for file generation. However, for one-off sharing by directly selecting a format it is not necessary.

Agree on the logic as it is now, not on the names. The new `isUIReady` is indeed `isDocReady` and should be called that way. The other thing, until now called isDocReady, can be renamed `isReadyToExport` or something similar (honoring the name that it used to have before the refactor and giving a proper name of what it does).

**Review by pazos — COMMENTED** [review]
Except for renaming a couple of methods and removing the timestamp the rest looks quite good and it is almost ready for prime time :)

:+1:

File: plugins/exporter.koplugin/main.lua [hunk 1]
```diff
@@ -193,14 +200,29 @@ end
...
+        if v.shareable then
+            sharemenu[#sharemenu + 1] = { text = _("Share as " .. v.name), callback = function()
+                local clippings = self:getDocumentClippings()
+                local document
+                for _, notes in pairs(clippings) do
+                    document = notes or {}
+                end
+
+                if #document > 0 then
+                    v.timestamp = os.time()
```
[the remark this answers is not in the document]
Comment (uroybd) [reply plugins/exporter.koplugin/main.lua]: Okay, I'm removing timestamp generation here. Exporters can get a default one from the `getTimeStamp`.

**Comment by uroybd** [comment]
> Except for renaming a couple of methods and removing the timestamp the rest looks quite good and it is almost ready for prime time :)
> 
> +1

Renaming and other changes based on the review are done. :)

File: plugins/exporter.koplugin/main.lua [hunk 2]
```diff
@@ -127,8 +127,11 @@ function Exporter:isReady()
 end
 
 function Exporter:isDocReady()
-    local docless = self.ui == nil or self.ui.document == nil or self.view == nil
-    return not docless and self:isReady()
+    return not (self.ui == nil or self.ui.document == nil or self.view == nil)
```
[the remark this answers is not in the document]
Comment (uroybd) [reply plugins/exporter.koplugin/main.lua]: Done.

File: plugins/exporter.koplugin/base.lua [hunk 3]
```diff
@@ -23,6 +23,7 @@ function BaseExporter:_init()
     self.extension = self.extension or self.name
     self.is_remote = self.is_remote or false
     self.version = self.version or "1.0.0"
+    self.sharable = self.sharable or false
```
Comment (pazos) [inline plugins/exporter.koplugin/base.lua:26]: Remove this one, it is not used.

You call the property `shareable`elsewhere. And the property is not required by any of the BaseExporter methods.

So it is ok if is not defined here.

Comment (uroybd) [reply plugins/exporter.koplugin/base.lua]: Done.

**Review by pazos — APPROVED** [review]
LGTM

**Pull request merged**